### ISSUE TITLE
RFE: Add support for BOOTP client

### ISSUE THREAD

**Issue by colin-foster-in-advantage** [systemd/systemd#34885]
### Component

systemd-networkd

### Is your feature request related to a problem? Please describe

A BOOTP server can issue static IPs to a BOOTP client. This is by way of sending a REQUEST message that doesn't contain any Options. The response from the server is a REPLY, which includes an IP and subnet.

Notably, these BOOTP messages don't include Option 53: DHCP Message Type, and the exchange is only two messages (unlike the four message DHCP)

### Describe the solution you'd like

Add an option under the network file

```
[DHCPv4]
BOOTP=yes
```

This will cause the initial request to be sent without any options. When the reply is received it will take a path similar to that of `rapid_commit` where the IP can get applied without the rest of the DHCP state machine (REQUEST and ACKNOWLEDGE)

Example Request

![image](https://github.com/user-attachments/assets/324c3574-165a-444b-9c5c-de831ebedf1f)

Example Reply

![image](https://github.com/user-attachments/assets/2c41a47b-3e81-4ce8-a2bc-fc3766d6b3eb)

### Describe alternatives you've considered

The program `bootpc` was used to perform this. There is no alternative in `systemd` as far as I know.

**Comment by colin-foster-in-advantage** [comment]
I do have a proof-of-concept of this feature. If it is a feature that would be accepted, I'll polish it up, add documentation and testing, and start an RFC / PR.

**Comment by yuwata** [comment]
Please open as a PR, if you have something. It is not necessarily perfect, and then you can open a draft PR.

**Comment by poettering** [comment]
May I ask: why?

Isn't BOOTP obsolete?

**Comment by colin-foster-in-advantage** [comment]
That is a fair question.

Odd network configurations where there might be DHCP servers coming online at different times outside of our control.

BOOTP can offer a rogue-DHCP-Server avoidance strategy by simply not communicating with those servers. Partly because BOOTP is mostly obsolete as far as those servers are concerned.

**Comment by avramd** [comment]
We have a more pressing need for bootp client behavior from systemd-networkd. We have a customer prospect that has a legacy bootp installation which they require vendor products to integrate with it.

**Comment by avramd** [comment]
@evverx as requested, here are two tcpdump packet captures; Please note, neither of these are from a working system. We are currently waiting on the customer prospect to supply us with a reference bootp server implementation and configuration. We have been using `bootpd` in the interim. 

IMHO, neither `bootpc` nor `bootpd` should be considered adequate reference implementations of BOOTP. I have filed bugs in the debian package system for both. The underlying source has been EOL for ages, so I don't expect them to be implemented unless I do it myself. 

We have abandoned the `bootpc` client because it does not open a raw socket, and is therefore unable to consume the replies from `bootpd` b/c they are unicast to the address that it is assigning but is not yet configured. Thus the kernel drops these as non-local, and they don't make it to the socket. This is the bug I filed w/ bootpc. This is the first capture, and that is why the exchange is repeated. FWIW though, it is a "valid" BOOTP 2-message exchange. 

The second one may serve no purpose, save for to illustrate where we are at in development. This shows the exchange between `systemd-networkd (v252)` and `bootpd`. As you will see, `bootpd` is attempting to participate as a `DHCP` server when it sees DHCP Message Type, by doing a 4-message exchange. However it is not providing a lease - that is why `networkd` is rejecting the DHCP_ACK.

Our hope is fairly modest here:

- if `BOOTP=yes`
  - Treat both DHCP_OFFER and DHCP_ACK packets that have no lease as if they contained an infinite lease
    - _Note: FallbackLeaseLifetimeSec=infinity only does this for DHCP_OFFER packets_
  - Accept BOOTP_REPLY messages (i.e. those with no option 53)
    - treat them as if they had contained an infinite lease
    - process them like RapidCommit DHCP_ACKs even when option 80 is not present
  - Do not error-out when option 54 is missing (DHCP Server Identifier)
    - it's fine to use `siaddr` in this case. I believe the relevance of it in a BOOTP context is that when a device that has already been assigned an address reboots, it should first send a unicast to the previous bootp server rather than broadcast for anyone to reply. It is not important to us whether this actually happens or not.

[bootpc-bootpd-exchange.pcap.zip](https://github.com/user-attachments/files/17754419/bootpc-bootpd-exchange.pcap.zip)
[networkd-bootpd-exchange.pcap.zip](https://github.com/user-attachments/files/17754421/networkd-bootpd-exchange.pcap.zip)

**Comment by evverx** [comment]
> Treat both DHCP_OFFER and DHCP_ACK packets that have no lease as if they contained an infinite lease

I don't think I've seen RFCs where this is described. I don't think BOOTP servers are supposed to use those options either. Could you show where this behavior is mentioned? If it's some non-standard extension I don't think it should be implemented. I think https://github.com/systemd/systemd/pull/34888 is an attempt to arrive at something that can cover RFC-compliant BOOTP servers and @colin-foster-in-advantage's use case (which isn't compliant with the RFC technically but doesn't deviate from it significantly).

The other two points are compliant with the RFC.

**Comment by avramd** [comment]
Section 3 of RFC1534 states “A DHCP client MUST assume that an IP address returned in a message from a BOOTP server has an infinite lease.”

I believe this hinges on whether option 53 is present. Is that how we are distinguishing whether the reply came from ADH E.C. P or BOOTP server?

As long as we are omitting the DHCP message type in the initial request when BOOTP is enabled, then I agree.

My fundamental point is that any message deemed to be a BOOTP message should be treated as if an infinite lease had been supplied. 

That would be a more RFC – compliant way to address my request.

I am not in a position to look at the source at the moment, that may already be how it is implemented in the PR.

**Comment by evverx** [comment]
Got it. https://datatracker.ietf.org/doc/html/rfc1534#section-3 describes a DHCP client that can understand BOOTP replies. https://github.com/systemd/systemd/pull/34888 implements a BOOTP client (where the DHCP Message Type option is never included in its requests).

As far as I can remember I mentioned somewhere in https://github.com/systemd/systemd/pull/34888 that the fact that `BOOTP` is in the `[DHCPv4]` section can be confusing (because it can be interpreted differently) but the documentation already says that "a BOOTP Request is sent without any options by default" and in that mode DHCP Message Types effectively aren't recognized (even if they are included in replies for some reason) and the first reply is accepted.

**Comment by avramd** [comment]
I've been reading BOOTP & DHCP RFC's for several weeks now, and I'm finding it very difficult to navigate. There is no single RFC that says "This is what BOOTP is now," and obsoletes prior ones, or anything that says "This is what is DHCP but not BOOTP.

I don't think we can count on a lack of optional fields. Most of them have been formalized as part of BOOTP, just not required by it. It's not wrong for a BOOTP request to include some of those options, most of them are not "DHCP options" they are just "options" that BOOTP MAY support and DHCP MUST support.

I'm not sure of this yet, but I believe what distinguishes DHCP from BOOTP is:

- Option 53 "DHCP Message Type"
  - The associated requirement for either a 4-way message exchange or RapidCommit
  - The expectation that state is going to be maintained
  - The requirement to be prepared for further communication that depends on that state.

One thing that makes this difficult is that you can't really say "A DHCP client (or server) is not a BOOTP client (or server)," or vice versa. It comes down to which behavior patterns they support, and any one piece of software could support both.

I haven't finished RFC1534 yet, I hope it clarifies this. I will try to do so tomorrow.

I'm not concerned with the config section being named [DHCPv4]. Most of the code is the same either way, and DHCP is an extension of BOOTP - it makes sense that its a single client implementation in which BOOTP is simply a mode you can select.

One irony is that BOOTP appears to be entirely stateless. We're basically talking about faking a stateless protocol using a stateful implementation and feeding permanent state details to it. We could bypass the storing of state for BOOTP, but that would be more work & risk for no benefit.

**Comment by evverx** [comment]
The format of messages is different as well. In BOOTP the length of the `vend/options` field is 64 bytes (while in DHCP it's variable length). In the latest version of https://github.com/systemd/systemd/pull/34888 the `vend` field is padded automatically to be compliant with https://www.rfc-editor.org/rfc/rfc1542#section-3.5 and prevent various BOOTP servers from rejecting BOOTP requests networkd sends.

The meaning of the `siaddr` field is different. In BOOTP it's "server IP address returned in bootreply by server" while in DHCP it's "IP address of next server to use in bootstrap". This hasn't been addressed yet and that's why, for example, BOOTP replies sent by `dnsmasq` are rejected and interfaces go to the "failed" state with no way to recover automatically.

All in all I think https://github.com/systemd/systemd/pull/34888 is moving in the direction of uncovering various things and should arrive at something that can work with various BOOTP servers.

**Comment by evverx** [comment]
> This seems to suggest that the bootp option may deserve to be more than a boolean

Agreed. Though I wouldn't expect the hybrid mode to be implemented in https://github.com/systemd/systemd/pull/34888 (or anytime soon). I think it would be great to get at least BOOTP right first. It turned out to be surprisingly convoluted.

**Comment by avramd** [comment]
@evverx @colin-foster-in-advantage I'm hoping one of you is willing to give me a pointer. As a developer, Linux & systemd is a new domain for me. I followed the build instructions from the hacking guide, and I've built this branch and run its tests successfully. I'd like to actually try it out, but I have no idea how to do that. I haven't tried all the VS Code/debugger setup yet b/c all I want to do is drop it into a VM an just run it for real. Can you give me any pointers on how to do that?

I'm using UTM on a Mac, which is a frontend for QEMU. I'm running a VM from a pre-built image. In my specific case it probably won't work right now b/c my VM is Debian 12, which uses systemd v252 - I'm assuming this was branched off of v257rc1 or something close to that, and wouldn't be compatible with Debian 12. But if I can get a VM running w/ a newer version of systemd, would I be able to simply copy over some files from the branch build, and restart systemd, to test this branch out "for real?"

**Comment by evverx** [comment]
I think in terms of testing the BOOTP client it should be possible to log in to that VM and build an image there with that PR using mkosi (https://github.com/systemd/systemd/blob/main/docs/HACKING.md). Then it should be possible to launch a container with mkosi where that client is run. This way the VM and the container should be attached to the same local network (via the veth interface) and it should be possible to run `dnsmasq` (or any other bootp server) on the host the bootp client in the container can talk to.

Another option would probably be to build networkd in that VM and then run it from the build directory (but the problem is that the network interfaces on the Mac and in the VM should be set up somehow to make it possible for the BOOTP client to talk to some BOOTP server).

**Comment by colin-foster-in-advantage** [comment]
@avramd I would follow @evverx 's suggestions. I'm new to `systemd` myself. My testing was on actual hardware, as it was pretty easy to make changes to make changes, load an image, and test against our BOOTP server. I had a `bootpc` Wireshark in hand of what I needed to mimic, so the changes were actually pretty simple.

I found it useful to just run `gdb ./build/test-dhcp-client` once that test was put into place. From there, VSCode is an easy next step.

I'll jump back into #34888 soon. There's a lot of good discussion there and it looks like it has mostly settled, so I'll be able to make whatever changes are necessary.

**Comment by avramd** [comment]
Update: I've upgraded my VM's to Debian Unstable, they are on systemd v257rc3.

**Comment by DemiMarie** [comment]
> That is a fair question.
> 
> Odd network configurations where there might be DHCP servers coming online at different times outside of our control.
> 
> BOOTP can offer a rogue-DHCP-Server avoidance strategy by simply not communicating with those servers. Partly because BOOTP is mostly obsolete as far as those servers are concerned.

Aren’t such networks misconfigured?

Non-enterprise networks won’t have BOOTP anyway, and enterprise networks should be using switches that protect against rogue DHCP servers.

**Comment by evverx** [comment]
@avramd  Build failures with mkosi are generally considered bugs. Assuming it's reproducible with the main branch as well could you open a new issue and attach as many details as possible? `testlog.txt` should contain logs showing where `test-namespace` fails. (As far as I understand debian testing images are built in debian testing VMs run on a Mac with QEMU and I don't think this combination is tested regularly anywhere so it's possible it doesn't work very well)

**Comment by evverx** [comment]
@avramd I think containers run with `mkosi` would be the best option in terms of testing the bootp client because the network interface on the host and in the container should be set up automatically and the config in the container can be changed to turn on the BOOTP client (instead of the DHCP client running there automatically) and then run the BOOTP server on the host with all the interfaces up and running. If it isn't an option it would be necessary to tweak various setting to set up the interfaces.

> Can either of you give me any pointers on how to use image.raw with an alternate hypervisor? 

@DaanDeMeyer Could you take a look?

> can either of you point me to any docs on how to generate .deb packages for the development build such that I could use apt to install this on a device?

I don't think it's documented anywhere but mkosi should generate a bunch of `deb` files in the build directory and in theory it should be possible to install them (assuming images match the host). Generally `mkosi` is the official way to build systemd and various development images so if there are no docs or something doesn't work I think it should be OK to file issues so that it would be possible to make it easier to do all that. I'd wait for @DaanDeMeyer here.

**Comment by DaanDeMeyer** [comment]
The built images are not intended to be deployed on other hypervisors. We pass through a lot of settings via SMBIOS to make it work seemlessly that will all be missing if you try to boot it in UTM. You're in for a world of pain trying to make that work simply because no one ever did it before which means there'll be tons of stuff that doesn't work that you'll need to figure out.

If you don't have nested virtualization, the only thing I can recommend is `mkosi boot` to run a container instead of a virtual machine. If that doesn't work for testing this I'm afraid you'll have to find a system where you do have KVM available so you can use `mkosi qemu`.

Or you use the distribution packages produced by the mkosi build which can be found in the build/mkosi.output directory after a build and install those wherever you'd like to test the new feature.

All of this is documented in HACKING.md

**Comment by avramd** [comment]
@DaanDeMeyer thank you so much for your input. 

I've been using HACKING.md extensively. I'm finding it a very difficult document to follow. Is there any way I can provide feedback on that document, or would it be ok for me to propose some edits to it via a PR? 

I'd like to propose more and finer-grained sections & headings to make it easier to navigate, and some wording edits to some parts of it that are confusing and possibly misleading. The hope is to make it easier for someone from outside this community who is not already an expert in these tools to get up to speed and become a functional contributor here.

I will do my best to extract how to bring the functionality to another system via .deb packages. I can get KVM or use containers, but our next step will still be getting this onto a real machine. Ultimately I suspect we are going to have to back-port it to an older version of systemd, since Bookworm uses v252, and the next stable release of Debian probably either won't be here in time for us, or won't have this feature in it.

**Comment by DaanDeMeyer** [comment]
> @DaanDeMeyer thank you so much for your input. 
> 
> I've been using HACKING.md extensively. I'm finding it a very difficult document to follow. Is there any way I can provide feedback on that document, or would it be ok for me to propose some edits to it via a PR? 
> 
> I'd like to propose more and finer-grained sections & headings to make it easier to navigate, and some wording edits to some parts of it that are confusing and possibly misleading. The hope is to make it easier for someone from outside this community who is not already an expert in these tools to get up to speed and become a functional contributor here.
> 
> I will do my best to extract how to bring the functionality to another system via .deb packages. I can get KVM or use containers, but our next step will still be getting this onto a real machine. Ultimately I suspect we are going to have to back-port it to an older version of systemd, since Bookworm uses v252, and the next stable release of Debian probably either won't be here in time for us, or won't have this feature in it. 

Of course feel free to open a PR with the suggested improvements to the doc.

**Comment by evverx** [comment]
> some parts of it that are confusing and possibly misleading

I think it can be misleading partly because it covers use cases unrelated to testing networkd (and I should have mentioned that `mkosi boot -n` should be used along with `systemd-networkd` running on the host to let `nspawn` set up all the things in the container correctly. My bad).

Either way as far as I understand it's more or less settled and the packages generated by mkosi should suffice (assuming the interface in the VM is managed by networkd and set up in a way allowing the BOOTP client to talk to I assume an actual device with the BOOTP server on the same network where the Mac is).

**Comment by avramd** [comment]
@evverx @colin-foster-in-advantage I've booted into a containerized build. Thank you for the `-n` tip btw, that probably saved me a fair bit of time.

When I try to exercise the new `Bootp=yes` config option, I get `Unknown key 'Bootp' in section [DHCPv4], ignoring` in journalctl.

```
Nov 27 19:09:57 default systemd-networkd[248]: /etc/systemd/network/20-bootp.network:8: Unknown key 'Bootp' in section [DHCPv4], ignoring.
Nov 27 19:09:57 default systemd-networkd[248]: /etc/systemd/network/20-bootp.network:9: Failed to parse file: Invalid argument
```

That would seem to suggest that my container isn't running the version of systemd that I built. However it does match the current sha in my git repo.

Here's my network config file (in `/etc/systemd/network/20-bootp.network`):
```
[Match]
Name=host0@if3

[Network]
DHCP=yes

[DHCPv4]
Bootp=yes
```

Here's the version of systemd I'm seeing in the container:
```
root@default:~# systemctl --version
systemd 257 (257~rc2-ge487c8c)
```

And here's the current git sha in my project directory:
```
$ git log --oneline | head
e487c8ccb1 networkd: add ability to use BOOTP
```

LMK if you have any thoughts.

**Comment by evverx** [comment]
As far as I can remember in the latest version of the PR `Bootp` was renamed to `BOOTP` so `BOOTP=yes` should work.

**Comment by avramd** [comment]
@evverx it does work, I missed the rename. I haven't verified that the feature works yet, only that systemd accepts the config option.

**Comment by avramd** [comment]
@colin-foster-in-advantage thanks for checking in. Currently I'm unable to functionally complete a real-world test b/c a trick I need isn't working. When testing the Debian Bookworm version of systemd-networkd with `bootpd`, the latter rejects the requests coming from systemd out of hand b/c it decides the packets are "too small." I was getting around that by using `SendOption=` or `SendVendorOption=` to set an unused DHCP option for the sake of padding, to get `bootpd` to accept & process the request. (I've been using option code 254, and setting it to a string full of ascii U's so it's easy to see in a packet trace). 

Those options seem to be getting ignored on your branch. I haven't had a chance to do the same test on `main` yet. I also haven't had a chance to re-review your PR yet to see if you are explicitly skipping what these options are supposed to do.

**Comment by avramd** [comment]
@colin-foster-in-advantage cc: @evverx I've gotten a complete round-trip between the client and the server.

_Note: the issue with `SendOption=` not working was that I specified the interface name wrong under `[Match]` as `host0@if3` as it shows up in `ip addr show`, instead of just `host0`.)_

The exchange was not quite successful. I haven't had time to troubleshoot extensively, but wanted to offer a brief recap before I head out for the day. The client reported an error, and did not configure the interface. Here is the error: 
```
DHCP error: failed to get DHCP server IP address: No data available
```

This smells like the the implementation is expecting the DHCP server address to be specified in a DHCP option that is not present, rather than accept the BOOTP `siaddr` field as the server address.  

I've attached a packet capture. Here is journalctl output grepped for `DHCP` and not `DHCPv6`:

```
host0: DHCPv4 client: FREE
host0: DHCPv4 client: State changed: bound -> stopped
Registering bus object implementation for path=/org/freedesktop/network1/link iface=org.freedesktop.network1.DHCPServer
Registering bus object implementation for path=/org/freedesktop/network1/link iface=org.freedesktop.network1.DHCPv4Client
host0: Requested configuring of the DHCPv4 client.
host0: DHCPv4 client: STARTED on ifindex 2
host0: DHCPv4 client is configured, acquiring DHCPv4 lease.
host0: DHCPv4 client: DISCOVER
host0: DHCPv4 client: State changed: initialization -> selecting
host0: DHCPv4 client: Received message from RAW socket, processing.
host0: DHCPv4 client: ACK
host0: DHCPv4 client: State changed: selecting -> bound
host0: DHCP error: failed to get DHCP server IP address: No data available
host0: Acquiring DHCPv4 lease.
```
[add_bootp-sendoption.cap.zip](https://github.com/user-attachments/files/17951582/add_bootp-sendoption.cap.zip)

**Comment by avramd** [comment]
@colin-foster-in-advantage, @evverx I did get this to work. Systemd configured the container's interface using BOOTP=yes after I modified the code to use `siaddr` as `server_address` when `client->bootp` is true. 

 I'm attaching:
- journalctl output
- network config file
- bootptab
- bootpd server log
- a packet capture of the exchange. 
 
This is with 4 bytes of manual padding to make the bootpd server happy. Note that there is an extra first BOOTP request in the capture that is short, before the one that works - I don't know why, but it is consistent. 
[systemd-bootp-4byte-pad.txt](https://github.com/user-attachments/files/17953441/systemd-bootp-4byte-pad.txt)
[20-bootp.network.txt](https://github.com/user-attachments/files/17953442/20-bootp.network.txt)
[bootptab.txt](https://github.com/user-attachments/files/17953447/bootptab.txt)
[bootpd.txt](https://github.com/user-attachments/files/17953449/bootpd.txt)
[add_bootp-sendoption.cap.zip](https://github.com/user-attachments/files/17953450/add_bootp-sendoption.cap.zip)

**Comment by avramd** [comment]
Update: I restarted the container, and the short extra packet request went away. Now there are just two packets, one request & one reply. 

I also removed the forced padding setting, and it still works. Apparently the padding being done in the implementation is sufficient. 

I don't have push permission on either repository, so I'm attaching a patch of my change here. 
[server-siaddr.patch](https://github.com/user-attachments/files/17953540/server-siaddr.patch)

**Comment by evverx** [comment]
@avramd as far as I can remember the siaddr thing was brought up in https://github.com/systemd/systemd/pull/34888#discussion_r1835556705 (and since it can be triggered by BOOTP replies sent by `dnsmasq` (which is already integrated into the `networkd` test suite) it should be possible to add a test for that eventually to make sure it all works).

Generally the replies with no `siaddr` (but with the DHCP-specific server id option) @colin-foster-in-advantage mentioned still puzzle me and I'm not sure where they come from. They violate all the RFCs I've seen and it isn't clear to me whether `networkd` should accommodate them (for example I'm not sure what should happen if replies come with both the `siaddr` field set and the server identifier option).

I took a look at https://github.com/user-attachments/files/17953450/add_bootp-sendoption.cap.zip and as far as I can see the reply there look OK (in the RFC sense) and doesn't contain any DHCP-specific options.

Either way I think it would be great to move this discussion to the PR (closer to the code) because it's probably already hard to follow all the threads.
